lib/resolve/npm.js

```diff
@@ -22,29 +23,65 @@ module.exports = function resolveNpm (pkg, log) {
   // { raw: 'rimraf@2', scope: null, name: 'rimraf', rawSpec: '2' || '' }
   return Promise.resolve()
     .then(_ => toUri(pkg))
-    .then(url => {
-      return got.get(url).then(res => {
-        if (log) log('resolving')
-        return res.promise
-      })
-    })
+    .then(url => got.get(url).then(res => {
+      if (log) log('resolving')
+      return res.promise
+    }))
     .then(res => JSON.parse(res.body))
-    .then(res => {
-      return {
-        name: res.name,
-        fullname: '' + res.name.replace('/', '!') + '@' + res.version,
-        version: res.version, // used for displaying
-        dist: res.dist
-      }
-    })
-    .catch(errify)
+    .catch(err => err.statusCode === 404
+      ? getAllVersionsAndMatchOnClient(pkg)
+      : errify(err, pkg)
+    )
```

Comment:
aww yeah, i know using promises for pnpm wasn't a mistake, really makes things like this easier :D